t/001-wildcard-router.t

```diff
@@ -0,0 +1,113 @@
+use Test::Nginx::Socket::Lua 'no_plan';
+use Cwd qw(cwd);
+
+my $pwd = cwd();
+my $nginx = $ENV{TEST_NGINX_ROOT_PATH} || "$pwd";
+
+$ENV{TEST_NGINX_LUA_PATH} = "$nginx/src/?.lua;$nginx/src/?/init.lua;;";
+$ENV{TEST_NGINX_UPSTREAM_CONFIG} = "$nginx/nginx/upstream_wildcard.conf";
+$ENV{TEST_NGINX_DEFAULT_CONFIG} = "$nginx/nginx/default_wildcard.conf";
+$ENV{TEST_NGINX_SERVICE_CONFIG} = "$nginx/nginx/service_wildcard.conf";
+
+require("t/dns.pl");
+
+log_level('debug');
+repeat_each(2);
+no_root_location();
+run_tests();
+
+env_to_nginx(
+  'RESOLVER',
+  'API_HOST',
+  'ACCESS_TOKEN',
+  'APICAST_PORT'
+);
+
+__DATA__
+
+# === TEST 1: request though the wildcard router
+
+# --- main_config
+# env RESOLVER=127.0.0.1:1953;
+# env API_HOST=http://localhost:$TEST_NGINX_LUA_PATH;
+# env ACCESS_TOKEN=alaska-token;
+# --- http_config
+#   lua_package_path "$TEST_NGINX_LUA_PATH";
+#   include $TEST_NGINX_UPSTREAM_CONFIG;
+# --- config
+#   include $TEST_NGINX_DEFAULT_CONFIG;
+# --- request
+# GET /
+# --- more_headers
+# Host: foo-apicast-staging.example.com
+# --- udp_listen: 1953
+# --- udp_reply eval
+# $::dns->("apicast-staging", "127.0.0.1")
+# --- response_body chomp
+# credentials missing!
+# --- error_code: 401
+
+
+=== TEST 2: request though the wildcard router using the API
+
+--- main_config
+env RESOLVER=127.0.0.1:1953;
+env API_HOST=http://api.example.com:$TEST_NGINX_SERVER_PORT;
+env ACCESS_TOKEN=alaska-token;
+env APICAST_PORT=$TEST_NGINX_SERVER_PORT;
+--- http_config
+  lua_package_path "$TEST_NGINX_LUA_PATH";
+  include $TEST_NGINX_UPSTREAM_CONFIG;
+  server {
+    listen $TEST_NGINX_SERVER_PORT;
+    server_name foo-apicast-staging.example.com;
+    location / {
+      content_by_lua_block {
+        ngx.status = 200
```

Comment:
Guess in this case you'll actually have to listen on different port to fake APIcast. Otherwise the request is undisguisable from the original one. I'd suggest 1953 because we already use 1953 udp for DNS>